### PR TITLE
fix: improve pivot post-processing

### PR DESCRIPTION
**Pull request by betodealmeida**
### SUMMARY

This PR fixes many cases when applying the pivot table post-processing.

I reimplemented the algorithm from scratch to make sure it matches the JS logic, testing as many combinations as I could and comparing the Python results with the JS visualization.

For each of those cases I added a unit test showing the pivot. The resulting table in each of the unit tests matches the table from the visualization.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

N/A

### TESTING INSTRUCTIONS

N/A

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/charts/post_processing.py [hunk 1]
```diff
@@ -27,60 +27,151 @@
...
-        df = df.stack(0).unstack().reindex(level=-1, columns=metrics)
-
-    # flatten column names
-    df.columns = [
-        " ".join(str(name) for name in column) if isinstance(column, tuple) else column
-        for column in df.columns
-    ]
+    parts: List[Any] = list(label)
+    metric = parts[-1]
+    parts[-1] = metrics.index(metric)
```
Comment (eschutho) [inline superset/charts/post_processing.py:46]: would it be more or less pythonic to pop and then push instead?

**Review by eschutho — APPROVED** [review]
@betodealmeida walked me through this and it looks good. The tests are fantastic!

**Comment by henryyeh** [comment]
🏷️ 2021.31

**Comment by PowerPlop** [comment]
@betodealmeida  the new post processing seems to not export row headers when choosing the csv output?

postprocessing

**Comment by PowerPlop** [comment]
I see this one is tagged to be included in v1.3, but it seems to me that the missing headers is quite crucial for a csv export?

**Pull request merged**